### PR TITLE
Support for fnv1a_ch hashing

### PR DESCRIPTION
**Pull request by deniszh**
Adding support for FNV1a_ch hashing from https://github.com/grobian/carbon-c-relay for graphite-web. Except from its docs:
>"The fnv1a_ch cluster is a identical in behavior to carbon_ch, but it uses a different hash technique (FNV1a) which is faster but more importantly defined to get by a limitation of carbon_ch to use both host and port from the members. This is useful when multiple targets live on the same host just separated by port. The instance that original carbon uses to get around this can be set by appending it after the port, separated by an equals sign, e.g. 127.0.0.1:2006=a for instance a. When using the fnv1a_ch cluster, this instance overrides the hash key in use. This allows for many things, including masquerading old IP addresses, but mostly to make the hash key location to become agnostic of the (physical) location of that key. For example, usage like 10.0.0.1:2003=4d79d13554fa1301476c1f9fe968b0ac would allow to change port and/or ip address of the server that receives data for the instance key. Obviously, this way migration of data can be dealt with much more conveniently"

Jump_fnv1a_ch is better with metrics distribution, but it's more complicated, will add it later, if needed.

Fixing https://github.com/graphite-project/carbon/issues/602 and partially - https://github.com/graphite-project/graphite-web/issues/1645

### CODE REVIEW

**Comment by deniszh** [comment]
I can add more tests if needed - now I added only for changed functions. Tests data is taken from `carbon-c-relay` output, so, it should be compatible.

File: docs/config-local-settings.rst [hunk 1]
```diff
@@ -363,6 +363,11 @@ CARBONLINK_TIMEOUT
 
   Timeout for carbon-cache cache queries in seconds.
 
+CARBONLINK_HASHING_TYPE
+  `Default: carbon_ch`
+
+  Type of metric hashing function. Graphite-web has limited support of "fnv1a_ch" hashing from carbon-c-relay. Please note, that fnv1a_ch hashing will work here only if instance is defined.
```
Comment (obfuscurity) [inline docs/config-local-settings.rst]: Would be useful to link to the external project/docs. Also, what does `instance` refer to here?

Comment (obfuscurity) [inline docs/config-local-settings.rst]: Can we get an explicit list of supported values?

Comment (obfuscurity) [inline docs/config-local-settings.rst]: `s/support of/support for/`

File: webapp/graphite/local_settings.py.example [hunk 2]
```diff
@@ -285,6 +285,10 @@
 #CARBONLINK_HOSTS = ["127.0.0.1:7002:a", "127.0.0.1:7102:b", "127.0.0.1:7202:c"]
 #CARBONLINK_TIMEOUT = 1.0
 #CARBONLINK_RETRY_DELAY = 15 # Seconds to blacklist a failed remote server
+#
+# Graphite-web has limited support of "fnv1a_ch" hashing from carbon-c-relay
+# Please note, that fnv1a_ch hashing will work here only if instance is defined.
+#CARBONLINK_HASHING_TYPE = 'carbon_ch'
```
Comment (obfuscurity) [inline webapp/graphite/local_settings.py.example]: We should document all supported values.

File: webapp/graphite/render/hashing.py [hunk 3]
```diff
@@ -43,25 +61,32 @@ def compactHash(string):
...
+      small_hash = int(big_hash[:4], 16)
     return small_hash
 
   def add_node(self, key):
     self.nodes.add(key)
     self.nodes_len = len(self.nodes)
     for i in range(self.replica_count):
       replica_key = "%s:%d" % (key, i)
+      if self.hash_type == 'fnv1a_ch':
+        replica_key = "%d-%s" % (i, key[1])
```
Comment (obfuscurity) [inline webapp/graphite/render/hashing.py]: I would prefer an explicit `if else` here unless this is considered more pythonic.

**Comment by deniszh** [comment]
Tried to fix the docs, @obfuscurity 
Not sure if good now. Suggestions are welcome. :)

File: docs/config-local-settings.rst [hunk 4]
```diff
@@ -363,6 +363,14 @@ CARBONLINK_TIMEOUT
 
   Timeout for carbon-cache cache queries in seconds.
 
+CARBONLINK_HASHING_TYPE
+  `Default: carbon_ch`
+
+  `Possible values: carbon_ch, fnv1a_ch`
+
+  The default is "carbon_ch" - usual Graphite hashing function. Also, you can use "fnv1a_ch" hashing from `carbon-c-relay relay <https://github.com/grobian/carbon-c-relay>`_. Fnv1a_ch cluster uses instance name as the solely source of routing information. As described in "carbon-c-relay" documentation:
```
Comment (obfuscurity) [inline docs/config-local-settings.rst]: I would propose this instead, also removing the 2nd paragraph:

```
The default `carbon_ch` is Graphite's traditional consistent-hashing implementation. Alternatively, you can use the `fnv1a_ch`, which supports the Fowler–Noll–Vo hash function (FNV-1a) hash implementation offered by the `carbon-c-relay relay <https://github.com/grobian/carbon-c-relay>`_ project.
```

File: webapp/graphite/local_settings.py.example [hunk 5]
```diff
@@ -279,12 +279,32 @@
...
+# "fnv1a_ch" hashing from carbon-c-relay relay (https://github.com/grobian/carbon-c-relay)
+# Fnv1a_ch cluster uses instance name as the solely source of routing information.
+# As described in "carbon-c-relay" documentation: "The fnv1a_ch cluster is identical
+# in behavior to carbon_ch, but it uses a different hash technique (FNV1a) which
+# defined to get by a limitation of carbon_ch to use both host and port from the members.
+# When using the fnv1a_ch cluster, instance value overrides the hash key in use.
+# This allows for many things, including masquerading old IP addresses, but mostly
+# to make the hash key location to become agnostic of the (physical) location of that key.
+# For example, usage like 10.0.0.1:2003:4d79d13554fa1301476c1f9fe968b0ac would allow
+# to change port and/or IP address of the server that receives data for the instance key."
```
Comment (obfuscurity) [inline webapp/graphite/local_settings.py.example]: Let's take a similar approach here to what I suggested above.

**Comment by deniszh** [comment]
Done, @obfuscurity !

**Comment by obfuscurity** [comment]
@deniszh Did you see my full comment? I had replacement text with formatting for that paragraph.

**Comment by obfuscurity** [comment]
Hint: it was inside the blockquote. 😄

**Comment by deniszh** [comment]
Oops, was blind, sorry :)
Looks better now?

File: docs/config-local-settings.rst [hunk 6]
```diff
@@ -363,6 +363,12 @@ CARBONLINK_TIMEOUT
 
   Timeout for carbon-cache cache queries in seconds.
 
+CARBONLINK_HASHING_TYPE
+  `Default: carbon_ch`
+
+  `Possible values: carbon_ch, fnv1a_ch`
+
+  The default `carbon_ch` is Graphite's traditional consistent-hashing implementation. Alternatively, you can use the `fnv1a_ch`, which supports the Fowler–Noll–Vo hash function (FNV-1a) hash implementation offered by the `carbon-c-relay relay <https://github.com/grobian/carbon-c-relay>`_ project.
```
Comment (obfuscurity) [inline docs/config-local-settings.rst]: Sorry, one more fix (my fault)... s/you can use the/you can use/. Also in the settings example below. Thanks!

**Comment by deniszh** [comment]
NP, fixed!

**Comment by obfuscurity** [comment]
Thanks!

**Comment by deniszh** [comment]
Will squash and merge that EOD

**Pull request merged**